text/3371-cargo-target-dir-templates.md

```diff
@@ -0,0 +1,411 @@
+- It uses the workspace root dir and not manifest, which means a `targo script` would share cache between all the scripts (`cargo script`) in a directory, which may not be the desired effect.
+
+Some of those could be fixed of course, and I don't expect `cargo`'s `--target-dir` and `--manifest-path` to change or disappear anytime soon, but still, it could happen. An external tool like [`targo`][tg] will never be able to solve some of these or ensure forward compatibility as well as the solution proposed in this RFC.
+
+On the other hand, [`targo`][tg] is already here and working for at least one person, making it the most viable alternative for now.
+
+[source]: https://github.com/rust-lang/cargo/issues/11156#issuecomment-1285951209
+
+## Remapping
+
+[rust-lang/cargo#11156](https://github.com/rust-lang/cargo/issues/11156) was originally about remapping the target directory, not about having a central one but reading the issue, there seems to be no needs for more than the simple redefinition of the target directory proposed in this document. In the future, if `CARGO_TARGET_DIR_REMAP` is introduced, it could be used to be the prefix to the target directory like so:
+
+- Set `CARGO_TARGET_DIR_REMAP=/home/user/projects=/tmp/cargo-build`
+- Compile the crate under `/home/user/projects/foo/` **without** `CARGO_TARGET_DIR` set
+- The resulting target directory will be at `/tmp/cargo-build/foo/target`
+
+By making the priority order `CARGO_TARGET_DIR` > `CARGO_TARGET_DIR_REMAP` (when both are absolute paths) we would keep backward compatibility. Or we could disallow having the two set at once, so that they're alternatives and not ordered.
+
+When `CARGO_TARGET_DIR` is relative, the result could be `/tmp/cargo-build/foo/$CARGO_TARGET_DIR`.
+
+Overall, I feel remapping is much harder to implement well and can be added later without interfering with templates in `CARGO_TARGET_DIR` (and without this RFC interfering with remapping), though the design space is probably bigger than the one for this RFC.
+
+## Advantages of using a hash over regular subdirectories
+
+It's possible to achieve most of the same system proposed here by setting a value like `CARGO_TARGET_DIR="/base/dir/{manifest-path-dirs}"`, where a manifest in `/tmp/test1/test2/Cargo.toml` would resolve the build directory to `/base/dir/tmp/test1/test2/`, but most is not *all* of them:
+
+1. Predictable depth of subdirectories: even with a hash, depth can be unpredictable in some way since the untemplated part can be as long as necessary but by using a hash of predictable length, users can prepare for the limits imposed by their filesystem (for example on Windows, some forms of path are quite short): it would be surprising for a project to build when in `/` but not in `/tmp/some/what/deep/inside/folders` for example
```

Comment:
iirc its not depth of subdirectories but the character length of the path.

I'd recommend focusing this on that, e.g.

```
1. Hashes have a fixed length while `manifest-path-dirs` is dependent on the context, making it a hazard for cross-platform compatibility.  Say on Windows the target-dir is rooted in a user tmp dir and the manifest path is inside of the user documents.   Especially combined with corporate policies on names, those base paths alone can take up a good amount of the character budget without getting into project names, etc.
```

Reply:
Done in https://github.com/rust-lang/rfcs/commit/645d9c41cd581b3c9aaac7f78169cce391ad59dc